### DISCUSSION TITLE
addChatPattern : how to prevent infinite loop

### DISCUSSION THREAD

**Discussion by boly38** [PrismarineJS/mineflayer#3307] [Q&A] [answered]
Hi all,
thanks for this wonderful library 👍 

I had a question on **addChatPattern** ([doc](https://github.com/PrismarineJS/mineflayer/blob/master/docs/api.md#botaddchatpatternname-pattern-chatpatternoptions) - [+PR](https://github.com/PrismarineJS/mineflayer/pull/3304)) and would like to know how to prevent infinite loop ? 

a chat match a pattern, and the bot will anwer by producing a tchat that match a pattern too..

- here is an example on how I prevent it : https://gist.github.com/boly38/2d91ed5f17f9dcfaa559bfecb484eb1f (full example )

````javascript
    bot.addChatPattern(myEvent, /coucou/, {})
    bot.on(`chat:${myEvent}`, msg => {
        const botUsername = bot.username;
        if (msg[0].includes(botUsername)) {
            console.warn(`chat:${myEvent}: OH WAIT : dont trigger pattern on my own tchat`);
        } else {
            bot.chat("oh You say coucou")// including "coucou" here would cause infinite loop when trigger my own tchat is allowed
        }
```` 

- is there a better way to do that ?
- if no: do you plan to include an improvement in the library to protect the bot to produce infinite loop ?

in advance thanks

**Comment by amoraschi** [comment]
This is an issue with the regex, not an issue with mineflayer, I recommend using a page like https://regexr.com/ to test your regex patterns, also, you need capture groups for the event to return parameters.

**Reply by boly38** [reply]
first thanks for your answer @amoraschi 💚 

what are the requirements for the pattern argument?  because `/coucou/` is a valid regex that match any sentence that will include `coucou`. I was not aware that regex's group was mandatory. or at least not mandatory but change/improve handler args..

I even try to improve doc about this method (I added in my original post, the link to the doc, and to the related PR.
In the related sample, there is not always a group in the regex.

In my previous test I remark diff between handler args size depend of the test. So I'll now try to use group to see the diff. but anyway this could be great to add some details on usecase and/or improving related doc.

**Reply by boly38** [reply]
I updated [my gist](https://gist.github.com/boly38/2d91ed5f17f9dcfaa559bfecb484eb1f#file-addchattpattern-js-L47) 

- with a sample with pattern that embed one group : no change on the handler args
````
    const withOneGroupRegex = "withOneGroupRegex";
    bot.addChatPattern(withOneGroupRegex, /salut ([a-zA-Z]*)$/, {})
    bot.on(`chat:${withOneGroupRegex}`, (arg0,arg1,arg2,arg3,arg4,arg5,arg6) => {
        console.log(JSON.stringify({arg0,arg1,arg2,arg3,arg4,arg5,arg6}));
        // [t] salut toi
        // example : {"arg0":["<boly38> salut toi"]}
        const botUsername = bot.username;
        if (arg0[0].includes(botUsername)) {
            console.warn(`chat:${withOneGroupRegex}: OH WAIT : dont trigger pattern on my own tchat`);
        } else {
            bot.chat(`${withOneGroupRegex} oh You say >salut xxx`)// including pattern
        }
    });
````


- and with another sample with a pattern that embed one group + `parse:true` as options and I saw that with `parse:true` I got the first group capture in the first match group array.
````
    const withGroupRegexParseTrue = "withGroupRegexParseTrue";
    bot.addChatPattern(withGroupRegexParseTrue, /bonjour ([a-zA-Z]*)$/, {"parse": true})
    bot.on(`chat:${withGroupRegexParseTrue}`, (arg0,arg1,arg2,arg3,arg4,arg5,arg6) => {
        console.log(JSON.stringify({arg0,arg1,arg2,arg3,arg4,arg5,arg6}));
        // [t] bonjour toi
        // example :{"arg0":[["toi"]]}
    });
````

In fact this is well explain in the method doc in `parse` attribute. 
but all of this is not in the scope of my initial question on how to prevent loop I think.

NB: `repeat` option dont fit exactly to this question too because I'd like no trigger on myself and wont stop to reply after the first trigger.

In legacy `chat` event, in hander args, you've got username as first arg and this could (if custom event can produce this) fit my need here to prevent looping (even if library internal protection against loop may be a dedicated issue too)

**Comment by boly38** [comment, marked as answer]
The way to distinct author with addChatPatter is to use a regex and `parse:true` option. This way each group of regex is forwarded to the event listener and you could simply decide what to do with each message by author
Example
````
    addChatPattern(bot) {
        bot.addChatPattern('messageFromPlayer',
            /^<([^>]+)> (.*)$/, {"parse": true});
        bot.on('chat:messageFromPlayer', ([[user, userChat]]) => {
            console.log(`messageFromPlayer from ${user}: ${userChat}`);
        });
    }
````

Be warn that the regex and chat format may depend on minecraft server configuration
